### ISSUE TITLE
`reverse_proxy` with `HTTP_PROXY` and `NO_PROXY` environment variables

### ISSUE THREAD

**Issue by frozenice** [caddyserver/caddy#6030]
Sorry for the terseness, heading out, but wanted to post this before tomorrow. I can try to come up with a minimal reproducer, but maybe this is already enough for people who know their way around the dialer and proxy stuff in the Caddy codebase.

Caddy v2.7.6 h1:w0NymbG2m9PcvKWsrXO6EEkY9Ru4FJK8uQbYcev1p3A=

via Docker image `caddy:2.7.6-builder/alpine` with some plugins (which probably shouldn't affect things, they aren't used as an adapter or in the config here):

```
    --with github.com/abiosoft/caddy-yaml \
    --with github.com/mholt/caddy-l4 \
    --with github.com/ss098/certmagic-s3
```

Environment variables:

```
      HTTP_PROXY: "http://172.20.1.180:3128/"
      HTTPS_PROXY: "http://172.20.1.180:3128/"
      NO_PROXY: "localhost,127.0.0.1,192.168.0.0/16,172.16.0.0/12,10.0.0.0/8"
```

Relevant Caddyfile parts:

```
(reverse-proxy) {
  @websocket {
    header_regexp Connection "(?i)upgrade"
    header_regexp Upgrade "(?i)websocket"
  }

  handle @websocket {
    reverse_proxy http://{args[0]}
  }

  handle * {
    reverse_proxy h2c://{args[0]}
  }
}

domain {
   import reverse-proxy my-backend:8080
}
```

`my-backend` resolves to `10.234.1.4` inside the container, matching `NO_PROXY`.

When a non-WebSocket request comes in, the second `handle` block is triggered, which uses `h2c` as the upstream protocol. This works fine and results in a direct call to `my-backend`, not going through the `HTTP_PROXY`.

If the request is a WebSocket request, the first `handle` block is triggered, which uses `http` as the upstream protocol. This also results in a request to `my-backend`, but going through `HTTP_PROXY`, presumably ignoring `NO_PROXY`.

Also, when I change the second `handle` block to use `http` instead of `h2c`, it also results in a wrong backend call through the proxy.

(`h2c` in the first `handle` block results in `http2: invalid Upgrade request header: ["websocket"]`, which was the reason the split WS/non-WS in the first place)

**Comment by mohammed90** [comment]
I'm mobile right now, but note that this is not about the "http://" protocol part. It's probably the websockets part that trips it. The `HTTP_*` behavior comes from Go stdlib, so I'll have to check there. The question is: does Go stdlib adhere to the proxy env vars for websockets (which are TCP connections)?

**Comment by frozenice** [comment]
> It's probably the websockets part that trips it.

Doesn't look that way:

> Also, when I change the second handle block to use http instead of h2c, it also results in a wrong backend call through the proxy.

This is the non-WS block.

**Comment by frozenice** [comment]
TL;DR: misunterstanding + working as intended

Okay, so the first issue was my misunderstanding of the `NO_PROXY` env var. It normally doesn't work with IPs or CIDR Ranges, only hostnames pre-resolve. I copied it from somewhere where this was supported. What I needed to do was listing "my-backend" (and all other internal hosts) in `NO_PROXY`. A bit of pain, but okay.

Also `h2c` (HTTP/2) can't use normal HTTP Forward Proxies, which makes sense, when you think about it. :) SOCKS is possible, though, via the socks5:// scheme in `HTTP_PROXY` / `HTTPS_PROXY` (at least Go seems to support this, looking at [the code](https://tip.golang.org/src/vendor/golang.org/x/net/http/httpproxy/proxy.go#L146)). I haven't tested `reverse_proxy` with HTTP/2 and a socks5:// `HTTP(S)_PROXY` (not relevant for my case), but I assume it works.

So to summarize: it works as intended. h2(c) doesn't use http:// proxies.

One todo, which could come from all this, would be to add some notes in the docs (possibly at `reverse_proxy`) about `NO_PROXY` only accepting hostnames and HTTP/2 (HTTP/3) not supporting http(s):// proxies, but socks5:// proxies. Some further research / tests would be needed to cement theses claims.

Here is a reproducer and the test cases for posterity.

### Caddyfile

```
{
  debug
}

:9991 {
  reverse_proxy http://caddyserver.com
}

:9992 {
  reverse_proxy h2c://caddyserver.com
}
```

### Run via Docker

```shell
docker run --rm -it \
  -v "$PWD/Caddyfile:/etc/caddy/Caddyfile" \
  -p 9991:9991 \
  -p 9992:9992 \
  -e "HTTP_PROXY=127.0.0.1:9999" \
  -e "HTTPS_PROXY=127.0.0.1:9999" \
  -e "NO_PROXY=" \
  caddy:2.7.6-alpine
```

### http

```shell
curl -v http://localhost:9991
```

Proxy is used.

<details>

<summary>Logs</summary>

```
DEBUG   http.handlers.reverse_proxy     selected upstream       {
  "dial": "caddyserver.com:80",
  "total_upstreams": 1
}
DEBUG   http.handlers.reverse_proxy     upstream roundtrip      {
  "upstream": "caddyserver.com:80",
  "duration": 0.0006247,
  "request": {
    "remote_ip": "172.17.0.1",
    "remote_port": "57342",
    "client_ip": "172.17.0.1",
    "proto": "HTTP/1.1",
    "method": "GET",
    "host": "localhost:9991",
    "uri": "/",
    "headers": {
      "User-Agent": [
        "curl/7.81.0"
      ],
      "Accept": [
        "*/*"
      ],
      "X-Forwarded-For": [
        "172.17.0.1"
      ],
      "X-Forwarded-Proto": [
        "http"
      ],
      "X-Forwarded-Host": [
        "localhost:9991"
      ]
    }
  },
  "error": "proxyconnect tcp: dial tcp 127.0.0.1:9999: connect: connection refused"
}
ERROR   http.log.error  proxyconnect tcp: dial tcp 127.0.0.1:9999: connect: connection refused {
  "request": {
    "remote_ip": "172.17.0.1",
    "remote_port": "57342",
    "client_ip": "172.17.0.1",
    "proto": "HTTP/1.1",
    "method": "GET",
    "host": "localhost:9991",
    "uri": "/",
    "headers": {
      "User-Agent": [
        "curl/7.81.0"
      ],
      "Accept": [
        "*/*"
      ]
    }
  },
  "duration": 0.000847801,
  "status": 502,
  "err_id": "p1q18khwi",
  "err_trace": "reverseproxy.statusError (reverseproxy.go:1267)"
}
```

</details>

### h2c

```shell
curl -v http://localhost:9992
```

Proxy is not used.

<details>

<summary>Logs</summary>

```
DEBUG   http.handlers.reverse_proxy     selected upstream       {
  "dial": "caddyserver.com:80",
  "total_upstreams": 1
}
DEBUG   http.handlers.reverse_proxy     upstream roundtrip      {
  "upstream": "caddyserver.com:80",
  "duration": 0.397595998,
  "request": {
    "remote_ip": "172.17.0.1",
    "remote_port": "46486",
    "client_ip": "172.17.0.1",
    "proto": "HTTP/1.1",
    "method": "GET",
    "host": "localhost:9992",
    "uri": "/",
    "headers": {
      "User-Agent": [
        "curl/7.81.0"
      ],
      "Accept": [
        "*/*"
      ],
      "X-Forwarded-For": [
        "172.17.0.1"
      ],
      "X-Forwarded-Proto": [
        "http"
      ],
      "X-Forwarded-Host": [
        "localhost:9992"
      ]
    }
  },
  "error": "unexpected EOF"
}
ERROR   http.log.error  unexpected EOF  {
  "request": {
    "remote_ip": "172.17.0.1",
    "remote_port": "46486",
    "client_ip": "172.17.0.1",
    "proto": "HTTP/1.1",
    "method": "GET",
    "host": "localhost:9992",
    "uri": "/",
    "headers": {
      "User-Agent": [
        "curl/7.81.0"
      ],
      "Accept": [
        "*/*"
      ]
    }
  },
  "duration": 0.401046206,
  "status": 502,
  "err_id": "9748xdzjc",
  "err_trace": "reverseproxy.statusError (reverseproxy.go:1267)"
}
```

</details>

### http with `NO_PROXY=caddyserver.com`

Use `  -e "NO_PROXY=caddyserver.com" \` in the `docker run` command.

```shell
curl -v http://localhost:9991
```

Proxy is not used.

<details>

<summary>Logs</summary>

```
DEBUG   http.handlers.reverse_proxy     selected upstream       {
  "dial": "caddyserver.com:80",
  "total_upstreams": 1
}
DEBUG   http.handlers.reverse_proxy     upstream roundtrip      {
  "upstream": "caddyserver.com:80",
  "duration": 0.388812023,
  "request": {
    "remote_ip": "172.17.0.1",
    "remote_port": "54902",
    "client_ip": "172.17.0.1",
    "proto": "HTTP/1.1",
    "method": "GET",
    "host": "localhost:9991",
    "uri": "/",
    "headers": {
      "User-Agent": [
        "curl/7.81.0"
      ],
      "Accept": [
        "*/*"
      ],
      "X-Forwarded-For": [
        "172.17.0.1"
      ],
      "X-Forwarded-Proto": [
        "http"
      ],
      "X-Forwarded-Host": [
        "localhost:9991"
      ]
    }
  },
  "headers": {
    "Server": [
      "Caddy"
    ],
    "Date": [
      "Thu, 11 Jan 2024 13:58:17 GMT"
    ],
    "Content-Length": [
      "0"
    ],
    "Location": [
      "https://localhost/"
    ]
  },
  "status": 308
}
```

</details>

### h2c with `NO_PROXY=caddyserver.com`

Use `  -e "NO_PROXY=caddyserver.com" \` in the `docker run` command.

```shell
curl -v http://localhost:9992
```

Proxy is not used.

<details>

<summary>Logs</summary>

```
DEBUG   http.handlers.reverse_proxy     selected upstream       {
  "dial": "caddyserver.com:80",
  "total_upstreams": 1
}
DEBUG   http.handlers.reverse_proxy     upstream roundtrip      {
  "upstream": "caddyserver.com:80",
  "duration": 0.390563012,
  "request": {
    "remote_ip": "172.17.0.1",
    "remote_port": "40694",
    "client_ip": "172.17.0.1",
    "proto": "HTTP/1.1",
    "method": "GET",
    "host": "localhost:9992",
    "uri": "/",
    "headers": {
      "User-Agent": [
        "curl/7.81.0"
      ],
      "Accept": [
        "*/*"
      ],
      "X-Forwarded-For": [
        "172.17.0.1"
      ],
      "X-Forwarded-Proto": [
        "http"
      ],
      "X-Forwarded-Host": [
        "localhost:9992"
      ]
    }
  },
  "error": "unexpected EOF"
}
ERROR   http.log.error  unexpected EOF  {
  "request": {
    "remote_ip": "172.17.0.1",
    "remote_port": "40694",
    "client_ip": "172.17.0.1",
    "proto": "HTTP/1.1",
    "method": "GET",
    "host": "localhost:9992",
    "uri": "/",
    "headers": {
      "User-Agent": [
        "curl/7.81.0"
      ],
      "Accept": [
        "*/*"
      ]
    }
  },
  "duration": 0.391778014,
  "status": 502,
  "err_id": "y1vjqd29x",
  "err_trace": "reverseproxy.statusError (reverseproxy.go:1267)"
}
```

</details>
